pkg/controller/job/job_controller.go

```diff
@@ -808,15 +808,20 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 	var manageJobErr error
 
 	exceedsBackoffLimit := jobCtx.failed > *job.Spec.BackoffLimit
+	successCriteriaMetCondition, successCriteriaMet := hasSuccessCriteriaMet(&job)
 
-	if feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) {
+	// Transition of successCriteriaMet to Failed or FailureTarget is not allowed.
+	if !successCriteriaMet && feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) {
 		if failureTargetCondition := findConditionByType(job.Status.Conditions, batch.JobFailureTarget); failureTargetCondition != nil {
 			jobCtx.finishedCondition = newFailedConditionForFailureTarget(failureTargetCondition, jm.clock.Now())
 		} else if failJobMessage := getFailJobMessage(&job, pods); failJobMessage != nil {
 			// Prepare the interim FailureTarget condition to record the failure message before the finalizers (allowing removal of the pods) are removed.
 			jobCtx.finishedCondition = newCondition(batch.JobFailureTarget, v1.ConditionTrue, batch.JobReasonPodFailurePolicy, *failJobMessage, jm.clock.Now())
 		}
 	}
+	if successCriteriaMet {
+		jobCtx.finishedCondition = newCompleteConditionForSuccessCriteriaMet(successCriteriaMetCondition, jm.clock.Now())
```

Comment:
Can we just return the `SuccessCriteriaMet` condition here? I think we should delay generating `Complete` condition until there are no running pods, I don't think we can be sure of that at this point.

Reply:
Uhm, that sounds very reasonable. We can add `Complete` condition in `trackJobStatusAndRemoveFinalizers` when we can confirm if all pods are terminated.

Reply:
yeah, I think we should aim to follow the same approach as for FailureTarget

Reply:
@mimowo So, this is just a quetion. About FailureTarget, what is this purpose? https://github.com/kubernetes/kubernetes/blob/cc5362ebc17e1376fa79b510f7f354dbffe7f92e/pkg/controller/job/job_controller.go#L813-L814

This seems that the controller set a failed condition to finishedCondition if the job has FailureTarget condition.
This is similar to the current my successPolicy implementation.

Reply:
The purpose of failure target is to ensure we don't lose the information the pod matched failure policy when waiting for pods to be deleted. We want to set Failed only when pods are terminated. Otoh in case of controller łrestart we need to know that there was the pod which matched the policy. The pod could be garbage collected before all pods are terminated.

Reply:
Ah, it makes sense. Thank you for the clarifications :)

Reply:
Done.